### ISSUE TITLE
2.0 stable

### ISSUE THREAD

**Issue by bitinn** [bitinn/node-fetch#228]
@timothygu can we establish a roadmap for that, I see 2.0 alpha marked as main release again, are we going to release stable soon? are there ways to check adoption or do we have people actively testing alpha?

**Comment by bitinn** [comment]
(i am not against marking 2.0 alpha as main release, it will surely drive adoption, but a roadmap for what's to come might help)

**Comment by bitinn** [comment]
@TimothyGu just a quick notice, do you think it's time for another pre-release? I don't have the proper toolchain at the moment. If you got time this weekend, could you put out another pre-release? Otherwise I will get to it somehow.

**Comment by jkantr** [comment]
Hey is this possibly a thing I could assist with in the future? I've somewhat familiarized myself with Travis after today so I'm getting there... lol

**Comment by bitinn** [comment]
I pretty sure you can create a pre-release if you want:

https://github.com/bitinn/node-fetch/releases

To publish a npm release you have to go through me or Timothy for now :)

------

On the subject of stable 2.0 major release, there are still a few considerations, I will list them here:

- Do we wait for Fetch spec to settle or do we publish 2.0 in near future? I would suggest we go with the latter option, but then 3.0 seem unavoidable because [a number of key issues are still pending standard](https://github.com/bitinn/node-fetch/issues?q=is%3Aissue+is%3Aopen+label%3A%22standard+pending%22).

- 2.0 is a non-trivial upgrade (see [the upgrade guide](https://github.com/bitinn/node-fetch/blob/master/UPGRADE-GUIDE.md)), I would like to guarantee that once developers test their code and upgrade successfully, they don't have to worry about Fetch spec changes. Which means we should get the breaking changes in before putting out 2.0 stable.

  - Your PR should take care of #273.
  - I am not sure about #270 yet.
  - can #266 be fixed with a PR?

- So basically, if we release 2.0 stable then we are locked into non-breaking changes for quite sometime, are people happy with 2.0 as it is now? If we are, then I will greenlight it.

**Comment by jkantr** [comment]
I'll look into #266 and #270.. maybe tomorrow.

Those standard issues aren't going anywhere anytime soon I don't think. Especially the abort issue.. that's a mess since cancel-able promises are basically never happening. Fragments... fragments on Request and not on Response "seems right" but yeah, still vague I guess. We.. do strip them on response right? I mean you can't really have a fragment on response, no? Except in the Location header. I think the current behavior is likely to be what's settled on.

also #260 is basically not happening either.. the core nodejs/http team seems pretty set on the behavior of lower-casing being 'correct'.

So I don't think we have much we'd be waiting on in the near future.

**Comment by jkantr** [comment]
Sorry to doublepost. I also think the main README leaves a lot to be desired. Maybe we can work on a face lift for this to coincide with the 2.0 release? Also can leave a bold note at the top about the release with date and a link to the upgrade guide. Maybe I'll open a new issue asking for help or brainstorming on how to make the docs more user friendly.

**Comment by bitinn** [comment]
Yeah that would be great, readme is getting pretty long. I recall there are tricks to have collapsible readme on GitHub, though I haven't tried them myself.

**Comment by TimothyGu** [comment]
There's 

```html
<details><summary>One line summary</summary>

More

stuff

that's 

collapsed
</details>
```

<details><summary>One line summary</summary>

More

stuff

that's 

collapsed
</details>

Not sure how well it works on npm though.

**Comment by TimothyGu** [comment]
Also, I believe some details in the upgrade guide might need to be updated itself. I'm fairly occupied recently but I'll see what I can do.

**Comment by jkantr** [comment]
Honestly I think just having a nice contents list at the top with anchor links to requisite headings would make it 10x more usable (like most of the sections in the big block of example code could probably be broken up into its own section including a few more comments about each case). I think `winston` is a pretty good example of a 'long' readme that is still organized well and easy to use.

**Comment by wheresrhys** [comment]
Hello. Do you have any idea when you expect v2 to be out of alpha/beta. I'm working on v6 of fetch-mock, and would like to coordinate with upgrading the default version of node-fetch used if possible.

**Comment by bitinn** [comment]
It’s stable but won’t be out of alpha anytime soon due to frequent spec change, sometimes breaking.

(From my phone)

**Comment by bitinn** [comment]
I consider v2 to be the default choice as v1 only backport bugfixes, not spec change.

(From my phone)

**Comment by wheresrhys** [comment]
It can't be the default choice though when running `npm install node-fetch` will choose v1. 

What's the advantage of keeping it in alpha now, rather than just releasing a v3 when the spec changes mount up. It seems to me that the situation you have at the moment is a component that is unclear about whether it's stable at v2 or not

**Comment by bitinn** [comment]
> What's the advantage of keeping it in alpha now.

To reflect the fact that Fetch spec is not stable, and the Stream spec is also not stable, and promise cancellation was cancelled and now we are waiting [Fetch Abort to be implemented in browsers](https://github.com/w3c/web-platform-tests/pull/6484).

> Just releasing a v3 when the spec changes mount up

I am actually happy to do that. But as you can see resources are limited and someone need to step in and take ownership of v3 (actually we would like some help finishing up v2 if anyone is interested).

> Whether it's stable at v2 or not

- Stable as in fit for production.
- Not stable as in keeping semver promise.

PS: **I am happy to re-open the debate of whether we should wait for Abort in V2 or whether we should just call it a day and release V2.**

**Comment by bitinn** [comment]
To me, resolving these are a minimum requirement for V2:

- #95 (can decide to skip)
- #284 (need to provide a solution)
- #229 (linked to #284, need to provide a solution)
- Readme overhaul (see #367)

**Comment by lpinca** [comment]
> What's the advantage of keeping it in alpha now, rather than just releasing a v3 when the spec changes mount up.

+1 on having v2 released as stable. It seems that the above issues can be addressed in minor releases when ready.

**Comment by styfle** [comment]
+1 for 2.x today

I would think the abort controller would be an addition under a minor release and wouldn’t require a breaking change for a major release.

**Comment by bitinn** [comment]
Alright people I hear you, but if you can help with the other 3 issues by opening a PR it would make 2.0 stable came out much sooner.

Personally I seldom works on Nodejs nowadays because of my role and job changed. I am happy for someone to takeover v2 releases.

(From my phone)

**Comment by leoselig** [comment]
@bitinn 

Similar to @lpinca and @styfle I wonder if it would be reasonable to get v2 out "now".

Overall, v2 seems to address pressing issues in terms of spec-compliancy (e.g. [#127 – `Headers`-API]( https://github.com/bitinn/node-fetch/issues/127) or [`#51 – body.arrayBuffer()`](https://github.com/bitinn/node-fetch/issues/51).

On the other hand, the 3 mentioned issues do not seem mandatory for an upcoming major update:
- [#95 – Aborting a fetch](https://github.com/bitinn/node-fetch/issues/95): no finalized spec
- [#95 – Debugging Native HTTP request](https://github.com/bitinn/node-fetch/issues/284): library-specific enhancement
- [#229 – Add a `fetch` option to allow wrappers #229 ](https://github.com/bitinn/node-fetch/pull/229): library-specific enhancement

How do you feel about that? Am I missing any issues here?

**Comment by TimothyGu** [comment]
#95's spec has been finalized, but I agree it's not critical for v2.x and can totally be added incrementally. Same applies to the other two.

I also think we can do a stable v2.

**Comment by bitinn** [comment]
@TimothyGu then I think there is no roadblock on this, do you want to take the pleasure of releasing 2.0 stable? Any existing PR you want to merge before pushing it to npm?

Sorry for the lack of response as I am deep in a C# project land, but if you need to take a break or need further discussion you know how to reach me :)

Thx for driving v2 this far!

**Comment by TimothyGu** [comment]
I'll go through the PR list, and we should be able to have a v2.0.0 by the end of today!

**Comment by TimothyGu** [comment]
:disappointed: got tripped up by a million of other things... will try again this weekend

**Comment by TimothyGu** [comment]
v2.0.0 has been released! :tada:

Closing this issue, after a hair more than a year after it was created.
